walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import com.google.sps.Event;
+import com.google.sps.Events;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // If the request meeting is longer than the duration of one day, return an empty list.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // If there is no event on the calendar currently, return whole day as possible time slot.
+    if (events.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    // If no mandatory attendees or optional attendees, return whole day.
+    if (mandatoryAttendees.size() == 0 && optionalAttendees.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // Get unavailable time of mandatory attendees.
+    List<TimeRange> unavailableTimes = getUnavailableTimes(events, mandatoryAttendees);
+    // Merge unavailable intervals with overlaps.
+    List<TimeRange> mergedUnavailableTimes = mergeUnavailableTimes(unavailableTimes);
+    // Get available time slot base on merged unavailable time.
+    List<TimeRange> availableTimes = getAvailableTimes(mergedUnavailableTimes, request.getDuration());
+    return availableTimes;
+  }
+
+  private List<TimeRange> getUnavailableTimes(Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> unavailableTimes = new ArrayList<>();
+    for (Event e : events) {
+      Set<String> attendees = e.getAttendees();
+      for (String attendee : attendees) {
+        // If there is at least one mandatory attendee for the requested event, 
+        // mark the duration of this event as unavailable.
+        if (mandatoryAttendees.contains(attendee)) {
+          unavailableTimes.add(e.getWhen());
+          break;
+        }
+      }
+    }
+    return unavailableTimes;
+  }
+
+  private List<TimeRange> mergeUnavailableTimes(List<TimeRange> timeRanges) {
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+    List<TimeRange> mergedUnavailableTimes = new ArrayList<>();
+
+    for (TimeRange t : timeRanges) {
```

Comment:
To understand this algorithm, I think the important missing invariant (necessary to understand why the code is the way it is) is that `mergedUnavailableTimes` is a list of sorted, non-overlapping `TimeRange`s. Should that be documented?